banidb with pytest

MySmilingTurban opened the pull request.

MySmilingTurban committed:
Functions for Bani DB API py

MySmilingTurban committed:
Structure of JSON from GET in Python

MySmilingTurban committed:
Structure of JSON from GET in Python

MySmilingTurban committed:
Structure of JSON from GET in Python

Recently Updated

MySmilingTurban committed:
Update myfunctions.py

MySmilingTurban committed:
Merge branch 'main' into bani_db

MySmilingTurban committed:
All functions created !!

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Rename myfunctions.py to __main__.py

MySmilingTurban committed:
Delete structure.txt

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Delete bani_db directory

MySmilingTurban committed:
Merge branch 'main' into bani_db

MySmilingTurban committed:
Error removed

MySmilingTurban committed:
Update setup.py

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Delete bani_db directory

MySmilingTurban committed:
Create requirements.txt

MySmilingTurban committed:
Update requirements.txt

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Delete .github directory

MySmilingTurban committed:
Delete banidb.egg-info directory

MySmilingTurban committed:
Delete dist directory

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Delete banidb/__pycache__ directory

MySmilingTurban committed:
Update setup.py

MySmilingTurban committed:
Create python-package.yml

MySmilingTurban committed:
Create python-publish.yml

MySmilingTurban committed:
Merge pull request #1 from MySmilingTurban/bani_db

Bani db

MySmilingTurban committed:
Merge pull request #2 from MySmilingTurban/main

merging

MySmilingTurban commented:
Just recreated the workflows at my end and they all passed Awesomely.

MySmilingTurban committed:
Update requirements.txt

MySmilingTurban committed:
Merge pull request #3 from MySmilingTurban/main

Update requirements.txt

MySmilingTurban committed:
Revert "Update requirements.txt"

MySmilingTurban committed:
Merge pull request #4 from MySmilingTurban/revert-3-main

Revert "Update requirements.txt"

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,409 @@
+import requests
+import time
+import pickle
+y,m,d=time.localtime().tm_year,time.localtime().tm_mon,time.localtime().tm_mday
```
Please run a linter flake8 on this - it will fix the spaces and general formatting

MySmilingTurban replied:
Done👍🏼

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,409 @@
+        d.append(l)
+    return d    
+
+def save(dic):
```
In the zoom call we discussed that it would be nice if all the history related functionality is in its own module. But before doing that could you please investigate https://www.geeksforgeeks.org/python-functools-lru_cache/ 
If using that `function decorator` works, then maybe there is no need for the history module. If it feels overwhelming, pls. reach out to me.

MySmilingTurban replied:
@lbk003 ji,
I have used lru_cache everywhere possible but it just speeds up the results which have already been called once as they get stored in cache memory.
There is no way to access the passed parameters and data individually, as of what I know and researched so far.

KIndly, guide me through how to get the data from lru_cache

MySmilingTurban replied:
lru_cache of functools cannot give back the cached data,
so, I am recreating the lru_cache mechanism to access cached data using the OrderedDict structure.

MySmilingTurban replied:
Done👍

lbk003 commented on banidb/test_banidb.py:
```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+import pytest
+from banidb.banidb import *
+
```
I love how you put these tests in place with such little guidance.

lbk003 replied:
[This](https://realpython.com/python-mock-library/#:~:text=A%20mock%20object%20substitutes%20and,your%20code's%20behavior%20during%20testing.) explains why we use mocks for unit testing.

lbk003 commented on setup.py:
```diff
@@ -1,10 +1,12 @@
-)
\ No newline at end of file
+    url='https://github.com/KhalisFoundation/banidb-api-python',
+    keywords=['BaniDB','API','Amrit Keertan','Sundar Gutka','SikhiToTheMax','Gurbani','KhalisFoundation'],
```
also add `Sikh` `Guru Granth Sahib`

MySmilingTurban replied:
Done👍🏼

lbk003 commented on setup.py:
```diff
@@ -1,10 +1,12 @@
+    url='https://github.com/KhalisFoundation/banidb-api-python',
+    keywords=['BaniDB','API','Amrit Keertan','Sundar Gutka','SikhiToTheMax','Gurbani','KhalisFoundation'],
+    install_requires=['requests']
+)
```
Let's add the other fields as well `author_email` `download_url`  `classifiers`

MySmilingTurban replied:
all fields have been set.
download_url var created but the link will be assigned once it gets released.
**Provide author_email for the package @lbk003

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,409 @@
+    d=[]
+    for i in r['rows'][1:]:
+        if i['RaagID']==RaagID:
+            linkw='http://api.banidb.com/v2/writers'
```
Maybe refactor the base of the url  `https://api.bandidb.com` and version `v2` as global constants and construct the links within each method - that way in future it will be a single line change, should either of these change.

Also, let's use `https://` instead of `http://`

MySmilingTurban replied:
Done👍🏼

lbk003 reviewed (changes requested):
Overall - Great job ! 
Changes requested inline.

MySmilingTurban committed:
Create upcoming.md

MySmilingTurban commented:
All the requested changes are being worked upon one-by-one,
I will try to commit it to the PR asap

MySmilingTurban committed:
Update setup.py

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Delete history.dat

MySmilingTurban commented:
Tests with pytest are not yet updated.

lbk003 commented:
@MySmilingTurban - can you please fix the failing test, so we can merge your PR into `main`
Thanks!

MySmilingTurban committed:
Merge branch 'main' into bani_db

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Update __init__.py

MySmilingTurban commented:
Tests are being worked on...

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Basic tests which need to be mocked

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Updated mock tests

MySmilingTurban committed:
Updated data

MySmilingTurban committed:
History and Mocks

The whole mechanism for history management is set in history.py
Mocks of two functions have been made, waiting for review.

MySmilingTurban committed:
Fix class

lbk003 commented on setup.py:
```diff
@@ -1,10 +1,46 @@
+    name='banidb',
+    version='0.1',
+    license='MIT',
     description='BaniDB API for Python',
```
change to `Python API for BaniDB`

MySmilingTurban replied:
"Python Package for BaniDB API" will be better

MySmilingTurban replied:
?

MySmilingTurban replied:
> "Python Package for BaniDB API" will be better

?

lbk003 replied:
Sure

lbk003 commented on setup.py:
```diff
@@ -1,10 +1,46 @@
+    ],
+    install_requires=['requests'],
+    author_email='',
+    download_url='',
```
https://github.com/KhalisFoundation/banidb-api-python

MySmilingTurban replied:
download_url will be the one we get from the release.

lbk003 commented on banidb/history.py:
```diff
@@ -0,0 +1,60 @@
+            data = pickle.load(f)
+            self.cache = data[0]
+            self.capacity = data[1]
+        except (FileNotFoundError, EOFError):
```
I assume the EOFError that you are processing has something to do with the pickle load 
 as per this [SO post](https://stackoverflow.com/questions/24791987/why-do-i-get-pickle-eoferror-ran-out-of-input-reading-an-empty-file) 
Would you like to try to implement some of the recommendations from the post?

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+import time
+from .history import LRUCache
+from functools import lru_cache
+y = time.localtime().tm_year
```
add newline

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+cache = LRUCache()  # provide capacity integer
+
+
+def searchtype():
```
rename `searchtype` => `search_type`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+def search(query, searchtype=1, source='all', larivaar=False,
+           ang=None, raag=None, writer='all', page=1, results=None):
+    res = url+'/search/'+str(query)+'?'
+    urladd = {
```
rename `urladd` => `url_address`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+        verses = []
+        for verse in js['verses']:
+            verse_dict = {}
+            verse_dict['shabadId'] = verse['shabadId']
```
Can be merged into a single line 
``` verse_dict = {'shabadId': verse['shabadId']}```

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+
+def search(query, searchtype=1, source='all', larivaar=False,
+           ang=None, raag=None, writer='all', page=1, results=None):
+    res = url+'/search/'+str(query)+'?'
```
nit: Have you considered using formatted strings ? like
`res = f"{url}/search/{str(query)}?"`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+        'results': results
+        }
+    for i in urladd.keys():
+        res += i+'='+str(urladd[i])+'&'
```
formatted strings will make it more readable.

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+        res += i+'='+str(urladd[i])+'&'
+    results = {}
+    response = requests.get(res)
+    js = response.json()
```
Please consider renaming `js` to `json_blob` or something a bit more descriptive - so that the next developer who reads it understands  what's going on.

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    total_res = info['totalResults']
+    results['totalResults'] = total_res
+    current_page = info['pages']['page']
+    total_pages = info['pages']['totalPages']  # Total Pages
```
These are good readable variable names 👍🏽

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+
+
+def shabad(shabad_id, larivaar=False):
+    link = url+'/shabads/'+str(shabad_id)
```
formatted strings, please  😄

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    js = data.json()
+    info = js['shabadInfo']
+    shabad = {}
+    shabad['sourceUni'] = info['source']['unicode']
```
```shabad['source_uni']``` ... you get the idea

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    lines = []
+    for verse in js['verses']:
+        line = {}
+        line['verseID'] = verse['verseId']
```
Can be merged into a single line

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    if source_id == 'G':
+        if 0 >= ang_no or ang_no > 1430:
+            raise ValueError('Ang does not exist!')
+    link = url+'/angs/'+str(ang_no)+'/'+str(source_id)
```
as above

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+            raise ValueError('Ang does not exist!')
+    link = url+'/angs/'+str(ang_no)+'/'+str(source_id)
+    response = requests.get(link)
+    js = response.json()
```
as above

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+
+
+def hukamnama(year=y, month=m, day=d):
+    link = url+'/hukamnamas/'+str(year)+'/'+str(month)+'/'+str(day)
```
formatted strings

lbk003 commented on setup.py:
```diff
@@ -1,10 +1,46 @@
+        'Sundar Gutka'
+    ],
+    install_requires=['requests'],
+    author_email='',
```
`python@khalis.info`

lbk003 commented on banidb/test_banidb.py:
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
```
It's a nice practice to keep the tests in a separate folder
Would you mind moving this file to `tests/banidb/test_banidb.py`

lbk003 commented on banidb/history.py:
```diff
@@ -0,0 +1,60 @@
+import pickle
```
It is great that you have implemented a persistence LRUCache from scratch.

That being said, we can use, and should use a package if available.

Would you like to check this out ? 
https://pypi.org/project/requests-cache/ - does it meet your needs?

MySmilingTurban replied:
LRUCache from scratch is better, as we can manipulate our needs from the cache data as per our and user's convinience.

lbk003 commented on banidb/history.py:
```diff
@@ -0,0 +1,60 @@
+'''
+
+
+class LRUCache:
```
This is really a persistent cache - so we should name it either `History` or `PersistenceCache` 
It will be confusing to a new developer - why we just didn't use `@lru_cache`

MySmilingTurban replied:
@lru_cache stores the data which makes the request get handled faster, but the problem with it was that there wasn't anyone way to access that cache data to show up as history.
That's why recreated LRUCache from scratch

MySmilingTurban replied:
Also, it LRU format as it reorders the pairs of OrderedDict on every put

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    return results
+
+
+def koshsearch(query):
```
rename to `kosh_search`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    return results
+
+
+def koshword(word):
```
rename `kosh_word`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    return results
+
+
+def amritkeertanshabad(shabad_id):
```
rename to `amritkeertan_shabad`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    return akindices
+
+
+def amritkeertansearch(header_id):
```
rename to `amritkeertan_search`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+
+
+@lru_cache(maxsize=1)
+def amritkeertanindex():
```
rename to `amritkeertan_index`

lbk003 commented on banidb/banidb.py:
```diff
@@ -0,0 +1,506 @@
+    response = requests.get(link)
+    js = response.json()
+    chapters = {
+        'Rehat ID': rehat_id,
```
The keys need to have a consistent naming format - 
How about we stick to  `StudlyCaps` for class names and `snake_case` for everything else?

MySmilingTurban commented:
Today, the author_email and other fields will be added in setup.py

MySmilingTurban committed:
Added author_email in setup.py

MySmilingTurban commented:
Added author_email and Bug tracker url to setup.py

MySmilingTurban committed:
Create banidb.svg

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update banidb.svg

MySmilingTurban committed:
Update banidb.svg

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update banidb.svg

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Delete banidb.svg

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Rename banidbPy.svg to banidb.svg

lbk003 commented on banidb/test_banidb.py:
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
+from unittest.mock import Mock, patch
+import requests
```
See the attached file for sample tests for `shabad`

lbk003 replied:
```from banidb import banidb
import json


def test_shabad_ok_larivaar_false(requests_mock):
    shabad_url = f"{banidb.url}/shabads/1111"
    result = {"shabadInfo": {"shabadId": 1111,
                             "source": {"unicode": "source-unicode-ਸ੍ਰੀ",
                                        "english": "source-english",
                                        "pageNo": 111},
                             "writer": {"english": "writer-english"}
                             },

              "verses": [
                  {
                      "verseId": 11111,
                      "verse": {
                          "unicode": "verse-unicode-ਮਹਲਾ ॥"
                      },
                      "larivaar": {
                          "gurmukhi": "DnwsrImhlw5]",
                          "unicode": "ਧਨਾਸਰੀਮਹਲਾ੫॥"
                      },
                      "translation": {
                          "en": {
                              "bdb": "xlate-en-bdb",
                              "ms": "xlate-en-ms",
                              "ssk": "xlate-en-ssk"
                          },

                      },
                      "transliteration": {
                          "english": "transliteration-english",
                          "hindi": "transliteration-हिंदी",
                          "en": "transliteration-en",
                          "hi": "transliteration-हिं",
                          "ipa": "transliteration-ɐ ɑ ɒ ɓ ɔ ɕ ɖ",
                          "ur": "دھن-transliteration"
                      },

                  }]
              }

    # convert dict to string to json
    json_blob = json.loads(json.dumps(result))
    requests_mock.get(f"{shabad_url}", json=json_blob)

    # now run the function we are testing
    result = banidb.shabad(1111)

    assert result['sourceUni'] == 'source-unicode-ਸ੍ਰੀ'
    assert result['sourceEng'] == 'source-english'
    assert result['writer'] == 'writer-english'
    assert result['ang'] == 111
    assert result['verses'][0]['verseID'] == 11111
    assert result['verses'][0]['verse'] == 'verse-unicode-ਮਹਲਾ ॥'
    assert result['verses'][0]['steek'] == {
        'en': {'bdb': 'xlate-en-bdb', 'ms': 'xlate-en-ms',
               'ssk': 'xlate-en-ssk'}}
    assert result['verses'][0]['transliteration'] == {
        "english": "transliteration-english",
        "hindi": "transliteration-हिंदी",
        "en": "transliteration-en",
        "hi": "transliteration-हिं",
        "ipa": "transliteration-ɐ ɑ ɒ ɓ ɔ ɕ ɖ",
        "ur": "دھن-transliteration"
    }
    assert not 'larivaar' in result['verses'][0]


def test_shabad_ok_larivaar_true():
    # TODO
    pass


def test_shabad_not_ok_larivaar_false():
    # TODO
    pass


def test_shabad_ok_larivaar_true():
    # TODO
    pass

lbk003 replied:
requirements-test.txt 👇🏽 
```requests-mock==1.9.3```

MySmilingTurban replied:
I studied a bit about requests-mock.
It doesn't actually need to be imported or taken as a dependency as it works well with just pytest and requests

MySmilingTurban replied:
ok 👍

MySmilingTurban replied:
Thanks for the sample code.

lbk003 reviewed (changes requested):
Great work - keep persevering.

In addition to the comments inline
The following are also needed 
- A `requirements-test.txt`  with dependencies needed for the test files e.g. `requests-mock==1.9.3`
- A sample python program that calls all the functions - it will serve as sample code that people can refer to
- There seems to be mismatch between banidb.py and history.py  - there is error on `LRUCache.put`

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Delete banidb.svg

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Update README.md

MySmilingTurban committed:
Delete banidb_py.svg

MySmilingTurban commented:
While using `requests_cache` ,
data is getting cached but ain't accessible to be displayed to the user when asked.
Browsing to find possible ways for the same.

MySmilingTurban committed:
Requested changes implemented

History management improved! ( We are using our own Cache handler )
Tests are being worked on...

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
Add files via upload

MySmilingTurban committed:
setup.py fixes

MySmilingTurban committed:
fixes

MySmilingTurban committed:
Fixes

MySmilingTurban committed:
setup fixes

MySmilingTurban committed:
Delete test_banidb.py

MySmilingTurban committed:
test fixes

MySmilingTurban committed:
Delete banidb/tests directory

MySmilingTurban committed:
tests updated dir

MySmilingTurban committed:
Fixes

MySmilingTurban committed:
Test fixes

MySmilingTurban committed:
Fixes

MySmilingTurban commented:
@lbk003 , need help in solving error
`fixture 'requests_mock' not found`

MySmilingTurban committed:
Fixes

MySmilingTurban committed:
Update requirements-test.txt

MySmilingTurban committed:
Update requirements-test.txt

MySmilingTurban committed:
Fix for failing tests

MySmilingTurban committed:
Update __init__.py

MySmilingTurban committed:
Test fix

MySmilingTurban committed:
Delete context.py

MySmilingTurban committed:
Delete requirements-test.txt

MySmilingTurban committed:
Test fix

MySmilingTurban commented:
@lbk003  , 
Finally, all the tests passed and systemized, now we can smoothly add more tests.
Apologies for the excessive commits, because all the tests were passing on my desktop but failing on Github.
Chardikala 🔥

MySmilingTurban commented:
> Overall - Great job !
> Changes requested inline.

All requested changes implemented

lbk003 commented on tests/test_history.py:
```diff
@@ -0,0 +1,58 @@
+
+
+def test_clear():
+    cache.clear()
```
Shouldn't you first add something to the cache, clear it and then do the assert?

lbk003 replied:
Please keep the two tests separate. From a testing strategy perspective. 1 unit test should test a single piece of functionality.
Why ? because if this test fails, you will not know - without debugging - whether the cache.get failed, or the cache.check failed, or cache.clear failed.

lines 10-48 can be made a `pytest fixture` 
[See pytest fixture documentation](https://docs.pytest.org/en/latest/how-to/fixtures.html)

MySmilingTurban replied:
Done👍

lbk003 commented on upcoming.md:
```diff
@@ -0,0 +1,27 @@
+***Steps for BaniDB API py***
```
Please create github issues for each of these line items and remove the file.

lbk003 reviewed (changes requested):
3 small changes. 
I will approve and merge once they are done.

Thanks for your perseverance.

lbk003 commented on README.md:
```diff
@@ -1,2 +1,13 @@
+# BaniDB.[py](https://pypi.org/user/KhalisFoundation/)
+### _Python package for BaniDB API_
+
+[<img height=100 src='bdb.svg'>](http://banidb.com)
```
This does not display on my mac markdown reader.

This one does - 
`![](bdb.svg) `

MySmilingTurban committed:
Delete upcoming.md

MySmilingTurban committed:
.md fix

MySmilingTurban committed:
Test Fixes

MySmilingTurban commented:
Ready to merge 🤩👍

MySmilingTurban commented:
A few improvement fixes will be done, today.

MySmilingTurban committed:
Improvement Fix

- Added clear function, which facilitates the user in clearing cache data
- check() function added in our LRUCache class, used in place of functools' @lru_cache decorator, making the results run fast and directly from the cache.
- Query length for search function conditioned for >2, to make the results smooth.
- history function displays all recent searched data
- Data can be held in cache for searched shabad ids and other constants like sources, writers,etc.

*Ready for the final merge*

MySmilingTurban commented:
## Improvement Fix 🛠
- Added clear function, which facilitates the user in clearing cache data
- check() function added in our LRUCache class, used in place of functools' @lru_cache decorator, making the results run fast and directly from the cache.
- Query length for search function conditioned for >2, to make the results smooth.
- history function displays all recent searched data
- Data can be held in cache for searched shabad ids and other constants like sources, writers,etc.

**Ready for the final merge**🚀

MySmilingTurban committed:
Improved test_history

MySmilingTurban commented:
> ## Improvement Fix 🛠
> * Added clear function, which facilitates the user in clearing cache data
> * check() function added in our LRUCache class, used in place of functools' @lru_cache decorator, making the results run fast and directly from the cache.
> * Query length for search function conditioned for >2, to make the results smooth.
> * history function displays all recent searched data
> * Data can be held in cache for searched shabad ids and other constants like sources, writers,etc.
> * pytest.fixture improvement in tests for history.
> 
> **Ready for the final merge**🚀

- pytest.fixture improvement in tests for history.

lbk003 commented on tests/test_history.py:
```diff
@@ -0,0 +1,67 @@
+    assert checker[0] is True
+
+
+def test_clear():
```
Pass in the fixture. Then clear the cache and assert that it is empty

lbk003 commented on tests/test_history.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@pytest.fixture()
+def put():
```
okay to merge the PR, but this should be named something like `intialize_cache_with_shabad` instead of `put`
